Source LinkedIn Ads - add streams Creative Name

Zawar92 opened the pull request:
## What
In the connector [source-linkedin-ads](https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-linkedin-ads), While fetching creative reports we are unable to fetch creative name with creative id.

Related Issue: https://github.com/airbytehq/airbyte/issues/19666

## How
have added a stream `AdCreativeName` that fetches the `creative_name`. In the records I have also appended the `creative_id` from the parent stream **Creatives**.

## Recommended reading order
1. `x.java`
2. `y.python`

## 🚨 User Impact 🚨
Are there any breaking changes? What is the end result perceived by the user? If yes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [x] Code reviews completed
- [ ] Documentation updated
    - [] Connector's `README.md`
    - [] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

sherifnada reviewed (changes requested):
@Zawar92 could you fix the diff? it seems like unrelated changes got pulled in

Zawar92 commented:
@sherifnada Hi,

I these changes are from the rebase I took from the latest Airbyte main branch.

I will look into this or if they can be removed from your side than better.

jacqueskpoty commented:
@Zawar92 Thanks for your contribution. I will be reviewing your PR.

jacqueskpoty commented:
@Zawar92 Change the version of the connector source-linkedin-ads in the Dockerfile from 0.1.12 to 0.1.13

jacqueskpoty commented:
@Zawar92 Submit a successful acceptance test after completing the requested changes.

bazarnov commented on airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/schemas/ad_creative_name.json:
```diff
@@ -0,0 +1,76 @@
+  "type": "object",
+  "properties": {
+    "com.linkedin.ads.SponsoredUpdateCreativeVariables": {
+      "type": "object",
```
The type should be also `nullable` as of `type: ["null", "object"]`, please make sure all the types are declared that way.

Zawar92 replied:
Yes. I will do it.

Zawar92 replied:
Done.

bazarnov commented on airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/source.py:
```diff
@@ -336,6 +336,59 @@ class AdCreativeAnalytics(LinkedInAdsAnalyticsStream):
+
+    endpoint = "adCreativesV2"
+    search_param = "projection"
+    search_param_value = "(variables(data(*,com.linkedin.ads.SponsoredUpdateCreativeVariables(*,share~(subject,text(text),content(contentEntities(*(description,entityLocation,title))))))))"
```
What exactly are we looking for using this parameter? Can we simplify this? 
Do we update something with this part:
`..com.linkedin.ads.SponsoredUpdateCreativeVariables..` ?

Zawar92 replied:
In documentation this parameter requires such value.
[Here](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?tabs=curl&view=li-lms-unversioned)

bazarnov replied:
Didn't manage to find what you've meant, probably the value is comprehensive, please take a look again on it, probably we can unify/sumplify this value parameter based on Creatives stream logic

Zawar92 replied:
[Here](https://learn.microsoft.com/en-us/linkedin/marketing/integrations/ads/account-structure/create-and-manage-creatives?tabs=curl&view=li-lms-unversioned#get-a-creative-name)

bazarnov commented on airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/source.py:
```diff
@@ -336,6 +336,59 @@ class AdCreativeAnalytics(LinkedInAdsAnalyticsStream):
     pivot_by = "CREATIVE"
 
 
+class AdCreativeName(LinkedinAdsStream):
```
Can we re-use the `Creatives` implementation class to inherit from it and override only the necessary parts, like: 
`search_param_value` and `search_param`, the `parent_values_map` would have the value of `{"creative_id": "id"}` that way we will keep the right structure and will not miss any of the campaigns the user has.

Zawar92 replied:
Yes I think. This would be the best suitable way. I will look into it.

Zawar92 replied:
I tried it doing with inheriting only the parts `search_param_value`, `search_param`, `parent_values_map` but doesn't work

Zawar92 replied:
Done.

sajarin commented:
@Zawar92 can you address the above comments ASAP? Thanks!

sajarin commented:
@Zawar92 any updates here? We're trying to get this merged as soon as possible. Let me know if there are any blockers here.

Zawar92 commented:
@sajarin,
Yes I am trying to add the changes I will update it as soon as I am done.

bazarnov commented on airbyte-integrations/connectors/source-facebook-marketing/integration_tests/spec.json:
```diff
@@ -352,4 +352,4 @@
     }
   }
-}
+}
```
please exclude this file from your changes

natalyjazzviolin replied:
Seconding this, please remove the file from this PR.

bazarnov commented on airbyte-integrations/connectors/source-coingecko-coins/source_coingecko_coins/coingecko_coins.yaml:
```diff
@@ -67,4 +67,4 @@ streams:
   stream_names:
     - "market_chart"
-    - "history"
+    - "history"
```
please exclude this file from your changes

bazarnov commented on airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/source.py:
```diff
@@ -336,6 +336,63 @@ class AdCreativeAnalytics(LinkedInAdsAnalyticsStream):
 
+class AdCreativeName(Creatives):
+
+    endpoint = "adCreativesV2"
```
No need to re-declare the endpoint var here, since it's a part of Creatives class

bazarnov commented on airbyte-integrations/connectors/source-linkedin-ads/source_linkedin_ads/source.py:
```diff
@@ -336,6 +336,63 @@ class AdCreativeAnalytics(LinkedInAdsAnalyticsStream):
+
+    sync_mode = SyncMode.full_refresh
+
+    @property
+    def source_defined_cursor(self) -> bool:
+        return False
+
+    @property
+    def cursor_field(self) -> str:
+        return ""
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return {}
+
+    def get_updated_state(self, current_stream_state: MutableMapping[str, Any], latest_record: Mapping[str, Any]) -> Mapping[str, Any]:
+        return {}
+
+    def request_params(
+        self, stream_state: Mapping[str, Any], next_page_token: Mapping[str, Any] = None, **kwargs
+    ) -> MutableMapping[str, Any]:
+        return {self.search_param: self.search_param_value}
```
do we actually need these methods after we inherit them from the Creatives class?

Zawar92 replied:
Methods like `source_defined_cursor` and `cursor_field` can only be removed. Other methods
1. `next_page_token`
2. `get_updated_state`
3. `request_params`

Needs to to added, as their usage in `Creative` class is different from what we require for this stream.

bazarnov commented on airbyte-integrations/connectors/source-postgres/src/main/java/io/airbyte/integrations/source/postgres/PostgresSourceOperations.java:
```diff
@@ -603,4 +603,4 @@ public boolean isCursorType(final PostgresType type) {
   }
 
-}
+}
```
please remove this file from your changes

bazarnov commented on airbyte-integrations/connectors/source-surveymonkey/source_surveymonkey/spec.json:
```diff
@@ -143,4 +143,4 @@
     }
   }
-}
+}
```
please remove this file from your changes

bazarnov commented on airbyte-integrations/connectors/source-twilio-taskrouter/source_twilio_taskrouter/twilio_taskrouter.yaml:
```diff
@@ -73,4 +73,4 @@ streams:
 check:
   stream_names:
-    - "workspaces"
+    - "workspaces"
```
please remove this file from your changes

bazarnov reviewed (changes requested):
Please check your implementation for more comments, thank you.

Zawar92 commented:
Hi@bazarnov, I will update with changes.

marcosmarxm commented:
Hello :wave::skin-tone-2: and thank you for your contribution!

Airbyte has instituted a code freeze between 19 and 30 December, to make sure there are no disruptions during the holidays. 
Because of this, reviewing and merging your contribution may take longer than usual. 
We apologize for the delay, but we want everyone to have a quiet and happy holiday.

If you have any questions or need further clarification, please don't hesitate to ping via Slack.

natalyjazzviolin commented on airbyte-integrations/connectors/source-plausible/integration_tests/configured_catalog.json:
```diff
@@ -10,4 +10,4 @@
     }
   ]
-}
+}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-plausible/integration_tests/invalid_config.json:
```diff
@@ -2,4 +2,4 @@
   "site_id": "https://airbyte.com",
   "start_date": "20220101"
-}
+}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-plausible/integration_tests/sample_config.json:
```diff
@@ -2,4 +2,4 @@
   "site_id": "airbyte.com",
   "start_date": "2022-01-01"
-}
+}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-plausible/source_plausible/schemas/stats.json:
```diff
@@ -34,4 +34,4 @@
       "description": "The number of visits/sessions."
     }
   }
-}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-plausible/source_plausible/spec.yaml:
```diff
@@ -36,4 +36,4 @@ connectionSpecification:
       pattern: ^[0-9]{4}-[0-9]{2}-[0-9]{2}$
       examples:
-        - YYYY-MM-DD
+        - YYYY-MM-DD
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-twilio-taskrouter/source_twilio_taskrouter/manifest.yaml:
```diff
@@ -71,4 +72,4 @@ streams:
 check:
   stream_names:
-    - "workspaces"
+    - "workspaces"
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-twilio-taskrouter/source_twilio_taskrouter/schemas/workers.json:
```diff
@@ -1,47 +1,47 @@
+              "type": ["null", "object"]
+            }
+        }
+}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-twilio-taskrouter/source_twilio_taskrouter/schemas/workspaces.json:
```diff
@@ -1,49 +1,49 @@
+              "type": ["null", "object"]
+            }
+        }
+}
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin commented on airbyte-integrations/connectors/source-coingecko-coins/source_coingecko_coins/manifest.yaml:
```diff
@@ -64,4 +64,4 @@ streams:
   stream_names:
     - "market_chart"
-    - "history"
+    - "history"
```
Please remove the file from this PR as it is unrelated to the LinkedIn Ads source connector.

natalyjazzviolin reviewed (changes requested):
Please make sure to remove any files that are not related to the LinkedIn Ads connector from this PR. Also please run integration tests locally and add the output to the template. Thanks!

Zawar92 commented:
@natalyjazzviolin,

I removed these files previously but after taking the recent rebase from master they get added.
I rebased it as as suggested by airbyte team on slack.
I will remove them again.

Zawar92 commented:
@natalyjazzviolin 

During the **integration test** I am facing runtime time for the API.

`Response Code: 429, Response Text: {"serviceErrorCode":101,"message":"Resource level throttle APPLICATION_AND_MEMBER DAY limit for calls to this resource is reached.","status":429}`

As seen in this [issue](https://github.com/airbytehq/airbyte/issues/19128)

Should I add these changes.

Zawar92 commented:
@marcosmarxm I removed all the conflicting files, give by @natalyjazzviolin.

marcosmarxm commented:
@Zawar92 conflicts are still there.
<img width="841" alt="image" src="https://user-images.githubusercontent.com/5154322/226478354-a3d275d1-8750-4714-abe7-887af9572b01.png">

Zawar92 commented:
I will remove them early morning.

Zawar92 commented on airbyte-integrations/connectors/source-coingecko-coins/source_coingecko_coins/manifest.yaml:
@marcosmarxm  here

marcosmarxm replied:
You're removing the files but they exist: https://github.com/airbytehq/airbyte/tree/master/airbyte-integrations/connectors/source-coingecko-coins/source_coingecko_coins

Zawar92 commented:
@marcosmarxm I was told to remove these files by other developers you can see the above comments. I was told to remove them. If you say so I can rebase it with the master.

marcosmarxm commented:
> @marcosmarxm I was told to remove these files by other developers you can see the above comments. I was told to remove them. If you say so I can rebase it with the master.
> 
> [#19930 (comment)](https://github.com/airbytehq/airbyte/pull/19930#issuecomment-1446953617)

I think @natalyjazzviolin said something you can read literally. What it said is to remove the changes to those files that your PR was causing it not to remove the files completely. You can `git merge master -Xtheirs` will correct your branch.

marcosmarxm commented:
/test connector=connectors/source-linkedin-ads
